### ISSUE TITLE
Create Notebook to obtain past egauge readings in the DHHL project 

### ISSUE THREAD

**Issue by jygh98** [erdl/legacy-scrape-util#1]
Eileen pointed out readings are still missing for DHHL egauges. We were unsure how to execute Forrest script manually to retrieve past data, so I tried using the code **locally** to see how it could be done. [A page on the wiki was added documenting the process](https://github.com/erdl/scrape-util/wiki/Manually-Running-Scrape-Util-to-Retrieve-Past-Data).

**Comment by jygh98** [comment]
## Setup

After cloning the scrape-util locally, open up \tmp and create a new directory called projects.  Within projects, you will be copying any of the project folders into it as found on the server.  In our case for egauges, i went ahead and copied dhhl, marine-center-egauge, maui-egauge, and uhm-frog.


![dsda](https://user-images.githubusercontent.com/25787662/29390355-1e4598e2-8283-11e7-98ff-d0e73aa44a48.png)


And inside those project folders are the necessary toml files


## Manipulating scrape-util
To manually set the desired start and end time for when you want the scrape-util to run from, simply overwrite it here in  [/src/acquire/egauge.py](https://github.com/erdl/scrape-util/blob/master/src/acquire/egauge.py)

>       [65]    start[gid] = nonce.get(gid,init)
>       [66]    stop[gid] = mkstop(start[gid])

For example, from 8/14/2017 to 8/15/2017

>       [65]    start[gid] = 1502668800
>       [66]    stop[gid] = 1502755200

To query only specific egauges within a project such as in dhhl, leave only the desired egauges in the config.toml file.  For example, in /tmp/projects/dhhl/config.toml, by default this is how the gauges dictionary is set

>     [acquire.egauge]
>     gauges = { 'e34111' = 34111 , 'e34113' = 34113, 'e34104' = 34104, 'e34105' = 34105, 'e34106' = 34106, 'e34107' = 34107, 'e34108' = 34108, 'e34109' = 34109, 'e34110' = 34110, 'e790' = 790}

That is, when the scrape-util is run it will query all the egauges within gauges.  To query only egauge e34111, remove all other egauges except it

>     [acquire.egauge]
>     gauges = { 'e34111' = 34111}


In combination with the time setter, we are able to query specific egauges within a desired timeframe and the output will be saved as a csv.

**Comment by carlosviansi** [comment]
@jygh98 I think Eileen already downloaded all the data directly from the website, right? Do you know if she added to the database?

**Comment by jygh98** [comment]
I am not to sure if she did downloaded the data directly, however I think scrape-util might have already uploaded the missing data as it does keep note of when the last export time was.  

As for running scrape-util locally, I changed the code to be able to run on the command of 
 > scrape-util "project name" "start date" "end date"
to pull data only from the specified dates

**Comment by carlosviansi** [comment]
@jygh98 Can you make a pull request with the associated Python Notebook?

**Comment by jygh98** [comment]
```
import requests

uri = 'http://egauge{}.egaug.es/cgi-bin/egauge-show?c&C&m'

start = 1505351760

stop = 1505354656.1330376

gauge = 34111

params = {'t': int(start), 'f': int(stop)}

r = requests.get(uri.format(gauge),params=params)

#these following lines is used to manipulate the data into a raw format that is reformattable and was taken directly form scrape-util

rows = [[y for y in x.split(',')] for x in r.text.splitlines()]

headers = [h.replace('"','') for h in rows.pop(0)]

columns = list(zip(*rows))

print({h: list(map(float,columns[i])) for i,h in enumerate(headers)})
```

**Comment by carlosviansi** [comment]
One more golden fish teased out of the code ;-) Great work. I forgot to ask you over phone, did you figure out if there was any "special trick" Forrest was requesting the data to avoid timeouts different than how we were doing? 

I am passing the torch from you to me to create the Notebook for now too.

**Comment by jygh98** [comment]
I don't think I noticed anything special that he did that would avoid the timeouts, do you know what caused these timeouts?

**Comment by carlosviansi** [comment]
The timeout is caused by the API, not a bug on our end. I just wanted to know where in Forrest code what he does if he gets a timeout of the API, the same way we did via the browser. I can imagine as the simplest solution simply to abort the rest of the script and wait for the next hour cron job call, but there should still be some `try` `catch` somewhere to fail gracefully. Does this make sense?

**Comment by jygh98** [comment]
I think what you just said how you imagine the solution is that the script aborts the rest of the script and waits for the next hour cron job call is correct.  If you take a look at the egauge scraper on [line 76](https://github.com/erdl/legacy-scrape-util/blob/master/src/acquire/egauge.py#L76), if it is returned a 200 error the function [returns back an empty dictionary](https://github.com/erdl/legacy-scrape-util/blob/master/src/acquire/egauge.py#L81).  This is then passed to raw which is as followed

     for gid in gauges:
        print('querying egauge: {}...'.format(gid))
        raw = query(gauges[gid],starts[gid],stops[gid])
        if not raw: continue

But because raw is empty the for loop ends and nothing is returned as queried.

**Comment by carlosviansi** [comment]
Following up on what we discussed offline, what I was referring to was for [this line](https://github.com/erdl/legacy-scrape-util/blob/2c4a718e3697cfef84493cdb1245baeb0d3b4693/src/acquire/egauge.py#L73):

```
    uri = 'http://egauge{}.egaug.es/cgi-bin/egauge-show?c&C&m'
```
Why the `C`?

**Comment by jygh98** [comment]
So from the documentation the '&C' parameter tells us the following,

    Specifies that the returned data be delta-compressed. That is, after the first
    row of data, each subsequent row’s columns are expressed as a difference relative
    to the previous row’s column-values. When combined with the CSV (c)
    parameter, the first row is always omitted

So from quick googling of what delta compression is, it is simply a method that allows you to quickly and efficiently download file data when you already have a similar file.  This is done by only transferring the difference between old and new files during update process, so this method is quite popular for software companies when they are distributing updated versions of the system since the receiving end already has the pre-existing files.  For our case, instead of redownloading all data again everytime we pull from eguard, essentially we would only be updating from our original source file and adding anything new to it.

**Comment by carlosviansi** [comment]
I hear you on the general meaning of deltas (thats how git works) but how is Forrest handling this once the data is saved? 

I am also a bit confused on this `the first row is always omitted` since this applies to us as it uses a .csv. 

Does Forrest pulls the existing data and use some operation to merge it? 

Truth be told I am not sure if on that file where the unix timestamps are saved so the script knows when to re-start it saves `last timestamp`, or `last timestamp + 1`. If it saves `last timestamp`, and it loads `last timestamp`, does this means the `C` is to ensure there is no overlap on the last row that was saved, and which would be the first row of when it is reused to download new data, thus avoiding overlap?

Does the rest of the .csv rows using `C` look just like your normal .csv other than missing the first row of data?

**Comment by jygh98** [comment]
So i tried printing out the data with the &C parameter and without, and here it is a sample of what i see

    (with &C) :   'Other [kW]': [0.482783333, 0.47885, 0.478866667, 0.484566667, 0.5006,
    (w/o &C):     'Other [kWh]': [912.182832778, 912.174786389, 912.166805556

And all the other column values very much look like this, being a difference of the prior data.  As to how Forrest integrate this varied data, i think this is done through the fmt_query() function.

**Comment by carlosviansi** [comment]
This whole idea looks very confusing to me. Is this one row you are showing the exact same column? 

But the difference to what exactly on the prior data? The last row?

**Comment by jygh98** [comment]
This is one column, the 'Other [kW]', where i am comparing with the parameter included or not.  So for the entire csv file of the &C parameter on, only the timestamps is returned unchanged.  Everything other column are expressed as a difference from the previous column.  Here are some sample data of each column of the &C parameter included

    {'Date & Time': [1505354640.0, 1505354580.0, 1505354520.0, 1505354460.0, 1505354400.0,
    'Whole house [kW]': [3.305216667, 2.15795, 2.1222, 3.3065
    'ACC [kW]': [1.233916667, 1.250883333, 1.23565, 1.236966667, 1.236066667,
    'Dryer [kW]': [1.3486, 0.262833333, 0.264283333, 1.441616667,

As compared to the entire values returned if we leave out the &C parameter.

     {'Date & Time': [1505354640.0, 1505354580.0, 1505354520.0,
     'Whole house [kWh]': [3364.746644444, 3364.6915575, 3364.655591667,
     'ACC [kWh]': [1625.369332778, 1625.3487675, 1625.327919444,

**Comment by carlosviansi** [comment]
What about the size of the files with C and without? Does it make a huge difference? Any guesses if this may impact the odds we can pull data out? I will take a look at the function. Odd how I couldn't get it when I searched in the repo.

**Comment by jygh98** [comment]
Comparing the size of the two files, it is not that significantly different considering that most of the csv files that we do get is on average around 100-300kb if it runs on the hourly cronjob.

![image](https://user-images.githubusercontent.com/25787662/30692550-51abecac-9e5f-11e7-8973-79edbec6c57e.png)

The first csv is with &C parameter included, the second csv is without.

**Comment by carlosviansi** [comment]
Can never tell whether these code decisions were made to use everything that is best regardless of the lack of need and make maintenance a living hell or not. I found something about "delta-uncompress": 

https://stackoverflow.com/questions/37951414/decompress-delta-compressed-sequence-of-numbers-in-scala

But it is in scala, another programming language:

```
def deltaDecompression(compressed : Seq[Long]) = {
  var previous = 0L
  compressed.map(current => {
    previous += current
    previous
  })
}

assert(deltaDecompression(Seq(100,1,2,3)) == Seq(100,101,103,106))
``` 

I guess the concept is just so common no one tries to explain clearly how to apply it in practice?

**Comment by carlosviansi** [comment]
@ryantanaka I am "Ccin" Ryan on this since I am starting to suspect there is really no real need to use this compression value. Here is my current lead:

http://www.dspguide.com/ch27/4.htm

From this link, there are some nice images:

![image](https://user-images.githubusercontent.com/17270563/30692819-d919ec30-9e68-11e7-8f8b-b108b3450f71.png)

and a visual representation of this:

![image](https://user-images.githubusercontent.com/17270563/30692845-e9bdef8c-9e68-11e7-9b79-75d6717ad090.png)

This is not the same page suggesting this, and the definition of the api as noted above by @jygh98:

```
after the first
row of data, each subsequent row’s columns are expressed as a difference relative
to the previous row’s column-values
```

Does seem to reflect it. As the website suggests:

```
The key feature is that the delta encoded signal has a lower amplitude than the original signal. In other words, delta encoding has increased the probability that each sample's value will be near zero, and decreased the probability that it will be far from zero. This uneven probability is just the thing that Huffman encoding needs to operate.
```

I know what Huffman encoding is, and I am starting to be afraid this amplitude range being decreased was solely done for the usage of the `zip` function....for the sole purpose of show how cool it is to decrease the amplitude of a sensor signal to zip and, well, apparently lead to little to no data compression on something that is already small and will be stored in the database uncompressed. 

But I am wearing the hat of the devil's advocate here. Any other takes? If the API does behave in a way it can get data with `C` and doesn't without it I am happy to maintain this. But these `zip` function calls paired with this `C` insofar and the size of the files are still leaving me with no clue why is this worth it. 

Additionally, I am not following how delta-uncompressing following the image method can apply to the data as I am not getting the values without the `C`. @jygh98 , are you positive the order is the same with and without the `C`?

**Comment by ryantanaka** [comment]
### code that goes through fmt_query() , prints variable values along the way to see manipulations

```
import requests
import sys
import data_utils as du

# Split the egauge supplied sensor/column name
# into separate units and sensor id.
# Expects egauge column headers to be
# of the form `SensorName [unit]`.
def parse_sntxt(sensor):
    try:
        '''
        print("printing parse_sntxt() -> sensor -------------")
        print(sensor)
        print("")
        '''
        unit = sensor.split('[').pop().replace(']','').replace(' ','')
        name = sensor.split(' [').pop(0).lower()
        '''
        print("printing parse_syntxt() -> unite,name -------------")
        print("unit: {}".format(unit))
        print("name: {}".format(name))
        print("")
        '''
    except Exception as err:
        print('failed to parse: {}'.format(sensor))
        print('parse error: {}'.format(err))
        # if parse fails, lower case column name is used as
        # sensor id, and unit is listed as "undefined".
        unit = 'undefined'
        name = sensor.lower()
    return name,unit

def fmt_query(gauge_id,data):
    # get all the time stamps
    dtimes = data['Date & Time']

    # values is the same dict as data but without "Date & Time" key value pair
    values = {k: data[k] for k in data if k != 'Date & Time'}
    print("printing fmt_query() -> values -------------")
    print(values)
    print("")

    # create empty list
    formatted = []

    # for each header in the csv
    for sn in values:
        # name : header name, unit: metric such as kW
        name,unit = parse_sntxt(sn)

        # make row is a function, that generates a row as a namedtuple
        # such that the time and value fields are casted as floats..
        mkrow = du.row_generator(gauge_id,name,unit)
        print("printing fmt_query() -> mkrow -------------")
        print(mkrow)
        print("")

        formatted += [mkrow(t,v) for t,v in zip(dtimes,values[sn])]
        print("printing fmt_query() -> formatted -------------")
        print("sn currently at '{}'".format(sn))
        print(formatted)
        print("")

    return formatted

gauge = 34111   # gauge = 34211
uri = 'http://egauge{}.egaug.es/cgi-bin/egauge-show?c&m'
start = 1505351760
stop = 1505354656

params = {'t': int(start), 'f': int(stop)}
r = requests.get(uri.format(gauge),params=params)

print("printing r.text-----------------")
print(r.text) # <----- this is CSV format already
print("")

# a list of lists where each nested list represents a row in the csv
rows = [[y for y in x.split(',')] for x in r.text.splitlines()]

print("printing rows-------------------")
print(rows)
print("")

# pop the first sublist off of rows (this contains the headers) and remove the commas
headers = [h.replace('"','') for h in rows.pop(0)]

print("printing headers----------------")
print(headers)
print("")

# for an arbritrary number of rows (thats why * is needed) return a list of tuples such
# that each tuple represents a column (for each row, zip gets all the items in the same index
# in each row and places them into a tuple)
columns = list(zip(*rows))

print("printing columns----------------")
print(columns)
print("")

# data is a dictionary such that each key word corresponds to a header in the csv and the
# value corresponds to a list of all items in that column under the header
# note that all values are being turned in to floats
data = {h: list(map(float,columns[i])) for i,h in enumerate(headers)}

print("printing data-------------------")
print(data)
print("")

print("printing output from fmt_query(gauge, data)")
fmt_query(786, data)

```

- the output won't look the nicest but the comments will help explain  exactly what was done with the code
- no computations done here to recompute original values, look else where
-  `fmt_query()` was taken from [here](https://github.com/erdl/legacy-scrape-util/blob/2c4a718e3697cfef84493cdb1245baeb0d3b4693/src/acquire/egauge.py#L93)
- `parse_syntxt()` was taken from [here](https://github.com/erdl/legacy-scrape-util/blob/2c4a718e3697cfef84493cdb1245baeb0d3b4693/src/acquire/egauge.py#L110)
- `du.row_generator()` is from [here](https://github.com/erdl/legacy-scrape-util/blob/2c4a718e3697cfef84493cdb1245baeb0d3b4693/src/core/data_utils.py#L26), use this file in the same directory as the above code for everything to run properly

* this comment is a little messy, i'll get back to it when i have time...

**Comment by carlosviansi** [comment]
I get it. The next step is looking at the PR by Jason since there is code in there that also decompress the delta to see what is the next candidate function that contains the solution and understand it I guess: 

https://github.com/erdl/legacy-scrape-util/pull/6/files

**Comment by carlosviansi** [comment]
i created the notebook to facilitate the process: https://github.com/erdl/sensors/blob/a5e4bcebb632b0cb454cb26bcf5559cc8b828f34/egauge/Requesting%20raw%20data%20from%20eGauge%20API.ipynb

Notice I added a conversion from datetime to unix direct into the notebook down to the hour, so this scan give you a quick way to text other time windows too without relying on the website. 

@jygh98 I am using the same last few lines to output the final content, so if you need Eileen to get something inserted into the database, you can run this notebook and just adjust to the portion of Forrest script that adds to the database. I will be working on the "insert into database" portion afterward as it requires the whole purpose_id stuff Eileen created. 
 
Now back to the delta compression.

**Comment by carlosviansi** [comment]
@jygh98 @ryantanaka 

The TL;DR to this issue is: Forrest did not really implement anything for delta-compression in `scrape-util`, and stores delta-compressed data. This is because delta-compression actually has a meaning: It is **Power (kW)** readings, whereas not specifying compression (i.e. `C`) means **Energy (kWh)**. 

Since Power can be aggregated to Energy by summing its value over the time window of interest, but the other way around is not true unless Power is held constant, this is why we request compressed data. 

I noticed this when I stared at the **column** label being different units.

However, this doesn't answer why delta-compression doesn't work if the name is being referred. I tried doing a test: I changed the granularity from minute (parameter `m`) to hour (parameter `h`) and tried to see if the delta-compression also didn't work: It **worked**!

In seeing the problem was at the minute level, I tried dividing the power table at minute granularity by **60** and follow the delta-compression concept of adding to the energy table....and **it matched!**. 

To top it off, when showing it to @kathrynparadis she mentioned that **Eileen used some /60 transformations** in Tableau. Coincidentally this Friday, and for other reasons, I asked Eileen to send me some Tableau examples she is using the data for at **minute** granularity from Forrest code. Since I had gone through the coursework of Tableau, I knew where to look at for user-defined-formulas.

Well, here is what I found:

<img width="516" alt="screenshot 2017-09-24 23 04 03" src="https://user-images.githubusercontent.com/17270563/30800748-ba7ee6ba-a17c-11e7-88dd-d252a229d756.png">

In Tableau, her **Power** column is as-is from Forrest script, but her Energy column is derived from Power divided by **60**, which matches my 60 assumptions from above.

I passed all this discussion by extending the Notebook on `eGauge` in a new section. It still covers just the basics of getting data out on the first section, but it expands on comparing 4 tables with resolution on minute and hour for both power and energy. I also added a minimal discussion to the default output, which is XML (but we don't use that). 

https://github.com/erdl/sensors/blob/c5981c91fa0cd304b8bcd0a0833550febf6bdfc3/egauge/Requesting%20raw%20data%20from%20eGauge%20API.ipynb

I asked @kathrynparadis to ask Eileen why she divides by **60** tomorrow. I know eGauge has a panel to configure how the **current transducers** are mapped into a single variable, so the owner can label the output from the API. For example, if an AC requires 2 CT, on this special panel in eGauge on the browser, (you need to have login/pass), you can create an equation summing both CT's outputs and call it "ACC" or whatever you want. 

This is why some egauges have some cryptic output, and others have a nice set of columns (this was configured by Eileen. 

You should also have her go over this with you. While I have access to the panel of egauges being on and off (also put on the notebook as a screenshot), I do not have the configuration. 

Hope this makes sense. 

If it is all clear, I will close this issue, and we can follow up on https://github.com/erdl/sensors/issues/4 on what Eileen explains on why she divides by 60 so I can confirm my hypothesis. 

With this, all that is left is #9 to wrap the eGauge and WebCTRL notebooks inserting data in the database and the nasty timestamp issue with `Arrow` and `Pendulum` libraries. See my last comment on [issue 3 on sensors](https://github.com/erdl/sensors/issues/3) for more details on this. 

p.s.: I've posted some dependency analysis plots to aid our function navigation on scrape-utils in the respective channels of Slack.
